src/util.ts

```diff
@@ -42,22 +42,24 @@ interface PackageJson {
   version: string;
 }
 
-export interface ClassOf<T, Config> {
+export interface Constructor<T, Config> {
   new(logger: Logger, config: Config): T;
   prototype: T;
   name: string;
 }
 
 /**
- * An class that provides access to a singleton.
+ * A class that provides access to a singleton.
+ * We assume that any such singleton is always constructed with two arguments:
+ * A logger and an arbitrary configuration object.
  * Instances of this type should only be constructed in module scope.
  */
 export class Singleton<T, Config> {
   private singleton: T|null = null;
 
-  constructor(private implementation: ClassOf<T, Config>) {}
+  constructor(private implementation: Constructor<T, Config>) {}
 
-  create(logger: Logger, config: Config&{forceNewAgent_: boolean}): T {
+  create(logger: Logger, config: Config&{forceNewAgent_?: boolean}): T {
```

Comment:
`Singleton` seems like a general purpose class, but this signature seems to specialize it for `TraceWriter`, so I am not sure if `Singleton` is the right name here. Do you really care about declaring the arguments here? Can you use `...args` and pass them onto the constructor below.

Reply:
We lose a significant amount of type-checking with `...args`. I don't see a need for constructors to be flexible; I believe that a logger and a configuration object should be universally applicable for anything that exists as a singleton (plugin loader, trace writer, possibly also cls in the future) within this library. Any data should be passed through configuration; any callbacks should arguably not be passed in through the constructor.

If it helps, I also wrote the assumption out in a comment that a singleton will need a logger and a config object.